File: python/rafter_cli/core/surface/differ.py
from __future__ import annotations

import json
from dataclasses import dataclass, field, replace
from typing import Callable, Mapping, Optional, Sequence

from .compare import danger_for, flip_order, lattice_compare, ordinal_compare, severity_for
from .kind_specs import KIND_SPECS
from .model import KindSpec, Property, SurfaceSeverity, Transition, Unanalyzed

SEVERITY_RANK = {"low": 0, "medium": 1, "high": 2, "critical": 3}


@dataclass
class DiffCoverage:
    unanalyzed: list[Unanalyzed] = field(default_factory=list)
    blocks_absence_proof: Optional[Callable[[Transition], bool]] = None


def _utf8(value: str) -> bytes:
    return value.encode("utf-8")


def _index_buckets(
    properties: Sequence[Property],
    side: str,
    specs: Mapping[str, KindSpec],
    coverage: DiffCoverage,
) -> dict[str, list[Property]]:
    groups: dict[str, list[Property]] = {}
    for prop in properties:
        groups.setdefault(prop.key, []).append(prop)
    indexed: dict[str, list[Property]] = {}
    for key, group in groups.items():
        try:
            spec = specs[group[0].kind]
        except KeyError as error:
            raise ValueError(f"missing kind spec for {group[0].kind}") from error
        if len(group) == 1 or spec.key_may_repeat:
            indexed[key] = group
            continue
        for file in sorted({prop.evidence.file for prop in group}, key=_utf8):
            coverage.unanalyzed.append(
                Unanalyzed(
                    file=file,
                    side=side,  # type: ignore[arg-type]
                    reason="parse_error",
                    detail=f"duplicate property key '{key}'",
                    changed=False,
                )
            )
    return indexed


def _level_vector(prop: Property, spec: KindSpec) -> str:
    return json.dumps(
        [prop.levels.get(axis.name) for axis in spec.axes],
        ensure_ascii=False,
        separators=(",", ":"),
    )


def _cancel_equal_vectors(
    base: Sequence[Property],
    head: Sequence[Property],
    spec: KindSpec,
) -> tuple[list[Property], list[Property]]:
    def sorted_by_vector(properties: Sequence[Property]) -> list[Property]:
        indexed = enumerate(properties)
        return [
            prop
            for _, prop in sorted(
                indexed,
                key=lambda item: (_utf8(_level_vector(item[1], spec)), item[0]),
            )
        ]

    sorted_base = sorted_by_vector(base)
    sorted_head = sorted_by_vector(head)
    left_base: list[Property] = []
    left_head: list[Property] = []
    base_index = 0
    head_index = 0
    while base_index < len(sorted_base) and head_index < len(sorted_head):
        base_vector = _utf8(_level_vector(sorted_base[base_index], spec))
        head_vector = _utf8(_level_vector(sorted_head[head_index], spec))
        if base_vector == head_vector:
            base_index += 1
            head_index += 1
        elif base_vector < head_vector:
            left_base.append(sorted_base[base_index])
            base_index += 1
        else:
            left_head.append(sorted_head[head_index])
            head_index += 1
    left_base.extend(sorted_base[base_index:])
    left_head.extend(sorted_head[head_index:])
    return left_base, left_head


def _classify(
    base: Optional[Property],
    head: Optional[Property],
    specs: Mapping[str, KindSpec],
    paired: bool = False,
    emit_unchanged: Optional[bool] = None,
) -> list[Transition]:
    endpoint = head if head is not None else base
    if endpoint is None:
        return []
    if base is not None and head is not None and base.kind != head.kind:
        raise ValueError(f"cannot compare {base.kind} with {head.kind}")
    try:
        spec = specs[endpoint.kind]
    except KeyError as error:
        raise ValueError(f"missing kind spec for {endpoint.kind}") from error
    if spec.comparator == "ordinal":
        order, axes = ordinal_compare(spec, base, head)
    else:
        order, axes = lattice_compare(spec, base, head)
    semantic_order = flip_order(order) if spec.invert_danger else order
    danger = danger_for(semantic_order)
    should_emit_unchanged = paired if emit_unchanged is None else emit_unchanged
    if danger == "unchanged" and not should_emit_unchanged:
        return []
    change = "added" if base is None else "removed" if head is None else "modified"
    key = endpoint.key
    if paired and base is not None and head is not None:
        key = min((base.key, head.key), key=_utf8)
    severity = severity_for(spec, base, head, danger, axes)
    return [
        Transition(
            kind=endpoint.kind,
            key=key,
            subject=endpoint.subject,
            label=endpoint.label,
            change=change,  # type: ignore[arg-type]
            danger=danger,
            severity=severity,
            axes=axes,
            from_level=axes[0].from_rank if len(axes) == 1 else None,
            to_level=axes[0].to_rank if len(axes) == 1 else None,
            confidence=endpoint.confidence,
            base_evidence=base.evidence if base is not None else None,
            head_evidence=head.evidence if head is not None else None,
            attrs=endpoint.attrs,
            paired=paired,
        )
    ]


def _relocation_pairs(
    base: Sequence[Property],
    head: Sequence[Property],
) -> tuple[list[tuple[Property, Property]], list[Property], list[Property]]:
    def relocation_key(prop: Property) -> Optional[str]:
        if prop.pairing_scope is None or prop.discriminator == "":
            return None
        return f"{prop.kind}\0{prop.pairing_scope}\0{prop.discriminator}"

    base_groups: dict[str, list[Property]] = {}
    head_groups: dict[str, list[Property]] = {}
    for prop in base:
        key = relocation_key(prop)
        if key is not None:
            base_groups.setdefault(key, []).append(prop)
    for prop in head:
        key = relocation_key(prop)
        if key is not None:
            head_groups.setdefault(key, []).append(prop)

    pairs: list[tuple[Property, Property]] = []
    used_base: set[int] = set()
    used_head: set[int] = set()
    for key in sorted(base_groups.keys() & head_groups.keys(), key=_utf8):
        base_group = base_groups[key]
        head_group = head_groups[key]
        if len(base_group) == 1 and len(head_group) == 1:
            pairs.append((base_group[0], head_group[0]))
            used_base.add(id(base_group[0]))
            used_head.add(id(head_group[0]))
    return (
        pairs,
        [prop for prop in base if id(prop) not in used_base],
        [prop for prop in head if id(prop) not in used_head],
    )


def _residual_pairs(
    base: Sequence[Property],
    head: Sequence[Property],
    specs: Mapping[str, KindSpec],
) -> tuple[list[tuple[Property, Property]], list[Property], list[Property]]:
    def scope_key(prop: Property) -> Optional[str]:
        spec = specs.get(prop.kind)
        if spec is None or not spec.allow_residual_pairing or prop.pairing_scope is None:
            return None
        return f"{prop.kind}\0{prop.pairing_scope}"

    base_groups: dict[str, list[Property]] = {}
    head_groups: dict[str, list[Property]] = {}
    for prop in base:
        scope = scope_key(prop)
        if scope is not None:
            base_groups.setdefault(scope, []).append(prop)
    for prop in head:
        scope = scope_key(prop)
        if scope is not None:
            head_groups.setdefault(scope, []).append(prop)

    pairs: list[tuple[Property, Property]] = []
    used_base: set[int] = set()
    used_head: set[int] = set()
    for scope in sorted(base_groups.keys() & head_groups.keys(), key=_utf8):
        base_group = base_groups[scope]
        head_group = head_groups[scope]
        if len(base_group) == 1 and len(head_group) == 1:
            pairs.append((base_group[0], head_group[0]))
            used_base.add(id(base_group[0]))
            used_head.add(id(head_group[0]))
    return (
        pairs,
        [prop for prop in base if id(prop) not in used_base],
        [prop for prop in head if id(prop) not in used_head],
    )


def _default_blocks_absence_proof(
    transition: Transition,
    unanalyzed: Sequence[Unanalyzed],
) -> bool:
    if transition.change == "added":
        return any(item.side == "base" for item in unanalyzed)
    if transition.change == "removed":
        return any(item.side == "head" for item in unanalyzed)
    return False


def diff_properties(
    base_properties: Sequence[Property],
    head_properties: Sequence[Property],
    specs: Sequence[KindSpec] = KIND_SPECS,
    coverage: Optional[DiffCoverage] = None,
) -> list[Transition]:
    active_coverage = coverage if coverage is not None else DiffCoverage()
    by_kind = {spec.kind: spec for spec in specs}
    base = _index_buckets(base_properties, "base", by_kind, active_coverage)
    head = _index_buckets(head_properties, "head", by_kind, active_coverage)
    output: list[Transition] = []
    unmatched_base: list[Property] = []
    unmatched_head: list[Property] = []

    for key in sorted(base.keys() & head.keys(), key=_utf8):
        base_group = base[key]
        head_group = head[key]
        try:
            spec = by_kind[base_group[0].kind]
        except KeyError as error:
            raise ValueError(f"missing kind spec for {base_group[0].kind}") from error
        if len(base_group) == 1 and len(head_group) == 1:
            output.extend(_classify(base_group[0], head_group[0], by_kind))
            continue
        left_base, left_head = _cancel_equal_vectors(base_group, head_group, spec)
        if len(left_base) == 1 and len(left_head) == 1:
            output.extend(_classify(left_base[0], left_head[0], by_kind))
        else:
            unmatched_base.extend(left_base)
            unmatched_head.extend(left_head)
    for key, group in base.items():
        if key not in head:
            unmatched_base.extend(group)
    for key, group in head.items():
        if key not in base:
            unmatched_head.extend(group)

    relocation_pairs, left_base, left_head = _relocation_pairs(
        unmatched_base, unmatched_head
    )
    for base_prop, head_prop in relocation_pairs:
        output.extend(
            _classify(base_prop, head_prop, by_kind, paired=True, emit_unchanged=False)
        )
    pairs, left_base, left_head = _residual_pairs(left_base, left_head, by_kind)
    for base_prop, head_prop in pairs:
        output.extend(_classify(base_prop, head_prop, by_kind, paired=True))
    for prop in left_base:
        output.extend(_classify(prop, None, by_kind))
    for prop in left_head:
        output.extend(_classify(None, prop, by_kind))

    for index, transition in enumerate(output):
        blocked = (
            active_coverage.blocks_absence_proof(transition)
            if active_coverage.blocks_absence_proof is not None
            else _default_blocks_absence_proof(transition, active_coverage.unanalyzed)
        )
        if blocked:
            output[index] = replace(transition, danger="unknown", severity=None)

    def sort_key(transition: Transition) -> tuple[int, bytes, bytes]:
        severity_rank = -1 if transition.severity is None else SEVERITY_RANK[transition.severity]
        return (-severity_rank, _utf8(transition.kind), _utf8(transition.key))

    return sorted(output, key=sort_key)

File: node/src/core/surface/differ.ts
import {
  dangerFor,
  flipOrder,
  latticeCompare,
  ordinalCompare,
  severityFor,
} from "./compare.js";
import { KIND_SPECS } from "./kind-specs.js";
import { canonicalJson } from "./serialize.js";
import type {
  KindSpec,
  Property,
  PropertyKind,
  SurfaceSeverity,
  Transition,
  Unanalyzed,
} from "./model.js";

export interface DiffCoverage {
  unanalyzed: Unanalyzed[];
  blocksAbsenceProof?: (transition: Transition) => boolean;
}

const SEVERITY_RANK: Readonly<Record<Exclude<SurfaceSeverity, null>, number>> = {
  low: 0,
  medium: 1,
  high: 2,
  critical: 3,
};

export function compareUtf8(left: string, right: string): number {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function indexBuckets(
  properties: readonly Property[],
  side: "base" | "head",
  specs: ReadonlyMap<PropertyKind, KindSpec>,
  coverage: DiffCoverage,
): Map<string, Property[]> {
  const groups = new Map<string, Property[]>();
  for (const property of properties) {
    const group = groups.get(property.key) ?? [];
    group.push(property);
    groups.set(property.key, group);
  }
  const indexed = new Map<string, Property[]>();
  for (const [key, group] of groups) {
    const spec = specs.get(group[0].kind);
    if (spec === undefined) throw new Error(`missing kind spec for ${group[0].kind}`);
    if (group.length === 1 || spec.keyMayRepeat) {
      indexed.set(key, group);
      continue;
    }
    for (const file of new Set(group.map((property) => property.evidence.file))) {
      coverage.unanalyzed.push({
        file,
        side,
        reason: "parse_error",
        detail: `duplicate property key '${key}'`,
        changed: false,
      });
    }
  }
  return indexed;
}

function specMap(specs: readonly KindSpec[]): Map<PropertyKind, KindSpec> {
  return new Map(specs.map((spec) => [spec.kind, spec]));
}

function levelVector(property: Property, spec: KindSpec): string {
  return JSON.stringify(spec.axes.map((axis) => property.levels[axis.name] ?? null));
}

/**
 * Content-derived ordering key for cancellation. Deliberately excludes evidence
 * (line numbers move when a document is reordered) and input position, so that
 * reordering semantically-unordered entries — IAM `Statement[]` order carries no
 * meaning — cannot change which property survives cancellation. Ties broken on
 * label then attrs, both of which are content.
 */
function contentSignature(property: Property): string {
  return canonicalJson([property.label, property.attrs]);
}

function cancelEqualVectors(
  base: readonly Property[],
  head: readonly Property[],
  spec: KindSpec,
): { leftBase: Property[]; leftHead: Property[] } {
  const sortByVector = (properties: readonly Property[]): Property[] => properties
    .map((property) => ({
      property,
      vector: levelVector(property, spec),
      signature: contentSignature(property),
    }))
    .sort((left, right) => compareUtf8(left.vector, right.vector)
      || compareUtf8(left.signature, right.signature))
    .map(({ property }) => property);
  const sortedBase = sortByVector(base);
  const sortedHead = sortByVector(head);
  const leftBase: Property[] = [];
  const leftHead: Property[] = [];
  let baseIndex = 0;
  let headIndex = 0;
  while (baseIndex < sortedBase.length && headIndex < sortedHead.length) {
    const baseVector = levelVector(sortedBase[baseIndex], spec);
    const headVector = levelVector(sortedHead[headIndex], spec);
    const order = compareUtf8(baseVector, headVector);
    if (order === 0) {
      baseIndex += 1;
      headIndex += 1;
    } else if (order < 0) {
      leftBase.push(sortedBase[baseIndex]);
      baseIndex += 1;
    } else {
      leftHead.push(sortedHead[headIndex]);
      headIndex += 1;
    }
  }
  leftBase.push(...sortedBase.slice(baseIndex));
  leftHead.push(...sortedHead.slice(headIndex));
  return { leftBase, leftHead };
}

function classify(
  base: Property | null,
  head: Property | null,
  specs: ReadonlyMap<PropertyKind, KindSpec>,
  paired = false,
  emitUnchanged = paired,
): Transition[] {
  const endpoint = head ?? base;
  if (endpoint === null) return [];
  if (base !== null && head !== null && base.kind !== head.kind) {
    throw new Error(`cannot compare ${base.kind} with ${head.kind}`);
  }
  const spec = specs.get(endpoint.kind);
  if (spec === undefined) throw new Error(`missing kind spec for ${endpoint.kind}`);
  const comparison = spec.comparator === "ordinal"
    ? ordinalCompare(spec, base, head)
    : latticeCompare(spec, base, head);
  const semanticOrder = spec.invertDanger ? flipOrder(comparison.order) : comparison.order;
  const danger = dangerFor(semanticOrder);
  if (danger === "unchanged" && !emitUnchanged) return [];
  const change = base === null ? "added" : head === null ? "removed" : "modified";
  const key = paired && base !== null && head !== null
    ? [base.key, head.key].sort(compareUtf8)[0]
    : endpoint.key;
  const severity = severityFor(spec, base, head, danger, comparison.axes);
  return [{
    kind: endpoint.kind,
    key,
    subject: endpoint.subject,
    label: endpoint.label,
    change,
    danger,
    severity,
    axes: comparison.axes,
    from: comparison.axes.length === 1 ? comparison.axes[0].from : null,
    to: comparison.axes.length === 1 ? comparison.axes[0].to : null,
    confidence: endpoint.confidence,
    baseEvidence: base?.evidence ?? null,
    headEvidence: head?.evidence ?? null,
    attrs: endpoint.attrs,
    paired,
  }];
}

function relocationPairs(
  base: readonly Property[],
  head: readonly Property[],
): { pairs: Array<[Property, Property]>; leftBase: Property[]; leftHead: Property[] } {
  const relocationKey = (property: Property): string | null => {
    if (
      property.pairingScope === null
      || property.discriminator === ""
    ) return null;
    return `${property.kind}\0${property.pairingScope}\0${property.discriminator}`;
  };
  const baseGroups = new Map<string, Property[]>();
  const headGroups = new Map<string, Property[]>();
  for (const property of base) {
    const key = relocationKey(property);
    if (key !== null) baseGroups.set(key, [...(baseGroups.get(key) ?? []), property]);
  }
  for (const property of head) {
    const key = relocationKey(property);
    if (key !== null) headGroups.set(key, [...(headGroups.get(key) ?? []), property]);
  }
  const usedBase = new Set<Property>();
  const usedHead = new Set<Property>();
  const pairs: Array<[Property, Property]> = [];
  const keys = [...baseGroups.keys()].filter((key) => headGroups.has(key)).sort(compareUtf8);
  for (const key of keys) {
    const baseGroup = baseGroups.get(key)!;
    const headGroup = headGroups.get(key)!;
    if (baseGroup.length === 1 && headGroup.length === 1) {
      pairs.push([baseGroup[0], headGroup[0]]);
      usedBase.add(baseGroup[0]);
      usedHead.add(headGroup[0]);
    }
  }
  return {
    pairs,
    leftBase: base.filter((property) => !usedBase.has(property)),
    leftHead: head.filter((property) => !usedHead.has(property)),
  };
}

function residualPairs(
  base: readonly Property[],
  head: readonly Property[],
  specs: ReadonlyMap<PropertyKind, KindSpec>,
): { pairs: Array<[Property, Property]>; leftBase: Property[]; leftHead: Property[] } {
  const scopeKey = (property: Property): string | null => {
    const spec = specs.get(property.kind);
    if (!spec?.allowResidualPairing || property.pairingScope === null) return null;
    return `${property.kind}\0${property.pairingScope}`;
  };
  const baseGroups = new Map<string, Property[]>();
  const headGroups = new Map<string, Property[]>();
  for (const property of base) {
    const key = scopeKey(property);
    if (key !== null) baseGroups.set(key, [...(baseGroups.get(key) ?? []), property]);
  }
  for (const property of head) {
    const key = scopeKey(property);
    if (key !== null) headGroups.set(key, [...(headGroups.get(key) ?? []), property]);
  }
  const usedBase = new Set<Property>();
  const usedHead = new Set<Property>();
  const pairs: Array<[Property, Property]> = [];
  const scopes = [...baseGroups.keys()].filter((key) => headGroups.has(key)).sort(compareUtf8);
  for (const scope of scopes) {
    const baseGroup = baseGroups.get(scope)!;
    const headGroup = headGroups.get(scope)!;
    if (baseGroup.length === 1 && headGroup.length === 1) {
      pairs.push([baseGroup[0], headGroup[0]]);
      usedBase.add(baseGroup[0]);
      usedHead.add(headGroup[0]);
    }
  }
  return {
    pairs,
    leftBase: base.filter((property) => !usedBase.has(property)),
    leftHead: head.filter((property) => !usedHead.has(property)),
  };
}

function defaultBlocksAbsenceProof(transition: Transition, unanalyzed: readonly Unanalyzed[]): boolean {
  if (transition.change === "added") return unanalyzed.some((item) => item.side === "base");
  if (transition.change === "removed") return unanalyzed.some((item) => item.side === "head");
  return false;
}

export function diffProperties(
  baseProperties: readonly Property[],
  headProperties: readonly Property[],
  specs: readonly KindSpec[] = KIND_SPECS,
  coverage: DiffCoverage = { unanalyzed: [] },
): Transition[] {
  const byKind = specMap(specs);
  const base = indexBuckets(baseProperties, "base", byKind, coverage);
  const head = indexBuckets(headProperties, "head", byKind, coverage);
  const out: Transition[] = [];
  const unmatchedBase: Property[] = [];
  const unmatchedHead: Property[] = [];

  const exactKeys = [...base.keys()].filter((key) => head.has(key)).sort(compareUtf8);
  for (const key of exactKeys) {
    const baseGroup = base.get(key)!;
    const headGroup = head.get(key)!;
    const spec = byKind.get(baseGroup[0].kind);
    if (spec === undefined) throw new Error(`missing kind spec for ${baseGroup[0].kind}`);
    if (baseGroup.length === 1 && headGroup.length === 1) {
      out.push(...classify(baseGroup[0], headGroup[0], byKind));
      continue;
    }
    const cancelled = cancelEqualVectors(baseGroup, headGroup, spec);
    if (cancelled.leftBase.length === 1 && cancelled.leftHead.length === 1) {
      // The key matched, but it does not discriminate within a repeated-key bucket:
      // which surviving statement corresponds to which is inferred, not proven, so
      // this carries paired=true exactly as Phase 2 residual pairing does. A 1:1
      // bucket is handled above as a true exact match. Cancellation has already
      // removed every equal-vector pair, so a residue pair never compares equal.
      out.push(...classify(cancelled.leftBase[0], cancelled.leftHead[0], byKind, true, false));
    } else {
      unmatchedBase.push(...cancelled.leftBase);
      unmatchedHead.push(...cancelled.leftHead);
    }
  }
  for (const [key, group] of base) {
    if (!head.has(key)) unmatchedBase.push(...group);
  }
  for (const [key, group] of head) {
    if (!base.has(key)) unmatchedHead.push(...group);
  }

  const relocated = relocationPairs(unmatchedBase, unmatchedHead);
  for (const [baseProperty, headProperty] of relocated.pairs) {
    out.push(...classify(baseProperty, headProperty, byKind, true, false));
  }
  const residual = residualPairs(relocated.leftBase, relocated.leftHead, byKind);
  for (const [baseProperty, headProperty] of residual.pairs) {
    out.push(...classify(baseProperty, headProperty, byKind, true));
  }
  for (const property of residual.leftBase) out.push(...classify(property, null, byKind));
  for (const property of residual.leftHead) out.push(...classify(null, property, byKind));

  for (const transition of out) {
    const blocked = coverage.blocksAbsenceProof?.(transition)
      ?? defaultBlocksAbsenceProof(transition, coverage.unanalyzed);
    if (blocked) {
      transition.danger = "unknown";
      transition.severity = null;
    }
  }

  return out.sort((left, right) => {
    const leftRank = left.severity === null ? -1 : SEVERITY_RANK[left.severity];
    const rightRank = right.severity === null ? -1 : SEVERITY_RANK[right.severity];
    if (leftRank !== rightRank) return rightRank - leftRank;
    const kindOrder = compareUtf8(left.kind, right.kind);
    return kindOrder !== 0 ? kindOrder : compareUtf8(left.key, right.key);
  });
}
